### PR TITLE
IGNITE-20030 Delete operation fails if PK is set on arbitrary columns.

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-20030

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/framework/TestBuilders.java [hunk 1]
```diff
@@ -669,6 +669,16 @@ public ChildT distribution(IgniteDistribution distribution) {
             return self();
         }
 
+        /** {@inheritDoc} */
+        @Override
+        public ChildT addKeyColumn(String name, NativeType type) {
+            columns.add(new ColumnDescriptorImpl(
+                    name, true, true, columns.size(), columns.size(), type, DefaultValueStrategy.DEFAULT_NULL, null
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/framework/TestBuilders.java:676]: why 'nullable' is true?

Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/framework/TestBuilders.java:676]: `The primary key constraint specifies that a column or columns of a table can contain only unique (non-duplicate), nonnull values. Technically, PRIMARY KEY is merely a combination of UNIQUE and NOT NULL.`

Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/framework/TestBuilders.java]: Thanks, fixed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java [hunk 2]
```diff
@@ -281,7 +281,7 @@ private <RowT> BinaryRowEx convertRow(RowT row, ExecutionContext<RowT> ectx, boo
                 continue;
             }
 
-            Object value = hnd.get(colDesc.logicalIndex(), row);
+            Object value = hnd.get(keyOnly ? colDesc.physicalIndex() : colDesc.logicalIndex(), row);
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java:284]: can we union these 2 loops : for (ColumnDescriptor colDesc : columnsOrderedByPhysSchema) { ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java [hunk 3]
```diff
@@ -126,4 +138,34 @@ private static PhysicalNode createAggregate(IgniteTableModify tableModify, RelOp
 
         return new IgniteProject(cluster, outTrait.replace(IgniteDistributions.single()), sumAgg, projections, convertedRowType);
     }
+
+    /**
+     * To perform delete we need row with key fields only.
+     * But input distribution contains the indexes of the key columns according to the schema (i.e. for the full row).
+     * This method aligns the keys to their indexes so that a row containing only the key fields can be read.
+     *
+     * <pre>
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java:147]: as discussed earlier - can you extend this java doc ?

Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java:147]: Expanded javadoc with more descriptive examples.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java (hunk 2 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java]: Thanks, removed first loop.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java [hunk 4]
```diff
@@ -281,29 +281,17 @@ public <RowT> CompletableFuture<?> deleteAll(
...
-
         RowAssembler rowAssembler = keyOnly ? RowAssembler.keyAssembler(schemaDescriptor) : new RowAssembler(schemaDescriptor);
 
         for (ColumnDescriptor colDesc : columnsOrderedByPhysSchema) {
             if (keyOnly && !colDesc.key()) {
                 continue;
             }
 
-            Object val = hnd.get(colDesc.logicalIndex(), row);
+            Object val = hnd.get(keyOnly ? colDesc.physicalIndex() : colDesc.logicalIndex(), row);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java:291]: this is not correct. Physical index specified order of columns in tuple that is stored in persistence, not the order after projection

**Review by korlov42 — CHANGES_REQUESTED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java [hunk 5]
```diff
@@ -126,4 +138,35 @@ private static PhysicalNode createAggregate(IgniteTableModify tableModify, RelOp
...
+     *  1. (a) is PK. Key fields row contains [a], we can read it using 0-index. No alignment required.
+     *  2. (b) is PK. Key fields row contains [b], distribution key contains logical index '1', which requires it to be "shifted" to '0'.
+     *  3. (b, d) is PK. Key fields row contains [b, d], distribution keys [1, 3] need to be shifted to [0, 1].
+     *  4. (d, b, c) is PK. Key fields row contains [d, b, c]. distribution keys [3, 1, 2] need to be shifted to [2, 0, 1].
+     * </pre>
+     *
+     * @param distribution Distribution specification.
+     * @return Distribution with adjusted keys.
+     */
+    private IgniteDistribution adjustDistributionKeysForDelete(IgniteDistribution distribution) {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java]: Thank you, I believe it's fixed now.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java (hunk 4 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java]: Thanks, fixed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/PlanUtils.java [hunk 6]
```diff
@@ -144,14 +144,33 @@ public static Mapping computeAggFieldMapping(List<ImmutableBitSet> groupingSets)
...
+     *  2 -> 0
+     *  3 -> 1
+     *  5 -> 2
+     *  7 -> 3
+     *  9 -> 4
+     * </pre>
+     *
+     * @param values Values to be mapped to its position in the ordered list.
+     */
+    public static Mapping sortedValuesIndexMapping(BitSet values) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/PlanUtils.java:165]: well, we already have `org.apache.ignite.internal.sql.engine.util.Commons#trimmingMapping`...

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java [hunk 7]
```diff
@@ -71,6 +74,11 @@ public final class UpdatableTableImpl implements UpdatableTable {
 
     private final List<ColumnDescriptor> columnsOrderedByPhysSchema;
 
+    private final List<ColumnDescriptor> keyColumnsOrderedByPhysSchema;
+
+    /** Mapping of column indexes for a row containing only primary key columns. */
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java:79]: let's add a few details describing WHY do we need such mapping

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java [hunk 8]
```diff
@@ -71,8 +76,27 @@ protected PhysicalNode convert(RelOptPlanner planner, RelMetadataQuery mq, Logic
...
+
+            for (int i = 0; i < igniteTable.descriptor().columnsCount(); i++) {
+                ColumnDescriptor column = igniteTable.descriptor().columnDescriptor(i);
+
+                if (column.key()) {
+                    keyFields.set(column.logicalIndex());
+                }
+            }
+
+            distribution = distribution.apply(PlanUtils.sortedValuesIndexMapping(keyFields));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java:95]: the same, let's use `org.apache.ignite.internal.sql.engine.util.Commons#trimmingMapping`

**Review by korlov42 — APPROVED** [review]
@xtern the patch looks good. I left a few minor comments though. It would be nice to have them fixed before merging

**Review by lowka — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java [hunk 9]
```diff
@@ -107,6 +119,24 @@ public UpdatableTableImpl(
         tmp.sort(Comparator.comparingInt(ColumnDescriptor::physicalIndex));
 
         columnsOrderedByPhysSchema = tmp;
+
+        int keyColumnsCount = schemaDescriptor.keyColumns().length();
+
+        List<ColumnDescriptor> keyCols = new ArrayList<>(keyColumnsCount);
+        List<Integer> keyLogicalIndexes = new ArrayList<>(keyColumnsCount);
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java:126]:
```suggestion
        int[] keyLogicalIndexes = new int[keyColumnsCount];
```

**Review by AMashenkov — CHANGES_REQUESTED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java [hunk 10]
```diff
@@ -71,8 +75,29 @@ protected PhysicalNode convert(RelOptPlanner planner, RelMetadataQuery mq, Logic
         IgniteTable igniteTable = relTable.unwrap(IgniteTable.class);
         assert igniteTable != null;
 
+        IgniteDistribution distribution = igniteTable.distribution();
+
+        if (rel.getOperation() == Operation.DELETE) {
+            // To perform the delete, we need a row with key fields only.
+            // Input distribution contains the indexes of the key columns according to the schema (i.e. for the full row).
+            // Here we adjusting distribution keys so that a row containing only the key fields can be read.
+            List<Integer> keyFields = new ArrayList<>();
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java:84]: Let's use IntList

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java [hunk 11]
```diff
@@ -107,6 +119,24 @@ public UpdatableTableImpl(
         tmp.sort(Comparator.comparingInt(ColumnDescriptor::physicalIndex));
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java]: sort removed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java [hunk 12]
```diff
@@ -107,6 +119,24 @@ public UpdatableTableImpl(
...
+        for (ColumnDescriptor colDesc : tmp) {
+            if (colDesc.key()) {
+                keyCols.add(colDesc);
+                keyLogicalIndexes.add(colDesc.logicalIndex());
+            }
+        }
+
+        keyColumnsOrderedByPhysSchema = keyCols;
+
+        ImmutableBitSet keysBitSet = ImmutableBitSet.of(keyLogicalIndexes);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java:139]: Done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java (hunk 10 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/TableModifyConverterRule.java]: Done

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**